---
title: "Module 7"
---

> Updated 2021/05/07

The Book of R: chapter 7.1, 7.2 (plots), ?system2

***

# Module 6 review

* Built in functions

    * ```
mysum = function(arg1) {
  return (sum(arg1))
  }
````

* Regular expressions
    * b[ce]+d
    * b[ce]*d
    
***

## Exercise: A useful function (10 minutes)

>Following on from the previous exercise where you found a string matching a regular expression...  
>Write a function that returns all the matches of a regular expression in a string  
>Hint: use substring (not substr) – why?  
>Test that the function works in the same way as in the previous exercise (by searching DNA for G..A)  

### Exercise solution (5 minutes)
```{r}
DNA <- "ACGATGATACTGATGACGGGCGCCA"  
REGEX <- "G..A"

# figuring things out ----
matches <- gregexpr(REGEX,DNA)[[1]]
substr(
  DNA,
  matches,
  matches + attr(matches,"match.length") - 1
)
substring(
  DNA,
  matches,
  matches + attr(matches,"match.length") - 1
)

# the function ----

MATCHER <- function(STRING,REGEX) {
  matches <- gregexpr(REGEX,DNA)[[1]]
  return(substring(
    DNA,
    matches,
    matches + attr(matches,"match.length") - 1
  ))
}

MATCHER(DNA,REGEX)
```

# Interfacing with external programs

* You can call/execute an external program from within your script
* Helps you automate certain tasks that would be difficult for you to do within R e.g. useful programs that are not available in R

```{r}
# system2() #calls an external command
system2("ls")  #unix/mac
system2("cmd.exe", input="dir")  #windows

```

## system2 function

```{r eval=F}
#returns stdout in variable v
v = system2("ls", stdout=T)
v

#passes arguments to the command (ls)
v = system2("ls", args="-la", stdout=T)
v

```

* What's wrong with system()?
    * It works, but is older and not recommended

## Exercise: external programs (10 minutes)

>Write a script to call a system command

* Mac/unix users
    * Run the “df –h” UNIX command from the shell – what does it do?
    * Call “df -h” from R to see how much space is left on your hard drive
    * Check your results against the “df -h” command you ran
* Windows users
    * Try to run windows "time" or "dir" commands
    * E.g. system2("cmd.exe", input="dir")
    * Can also try shell("dir")

### Exercise solution (5 minutes)
```{r}
# This was knit on a Windows OS:
try({
  system2("ls",stdout=T)
},silent=F)

system2(command="cmd.exe",input="dir",stdout=T)

shell("ls",shell="bash")

shell("df -h",shell="bash")
```


# Recipes
## Code examples and templates

* Recipes are standard way to do things
    * Look up the recipes in a cookbook 

![R recipe book](recipebookTOC.png)

## Example recipes

```{r,eval=FALSE,include=TRUE}
#Sorting a data frame
data_frame = data_frame[order(data_frame$key) , ]
#data_frame$key is the column to sort by
#Note: the sort() function only works for vectors, not data frames

#Sorting by two columns
data_frame = data_frame[order(data_frame$key1,data_frame$key2) , ] 
#The second key is used to break ties in the first

#Removing elements from a list
#Set the value to be NULL
```

## Stack Overflow
* User-contributed question and answer website – good free source of recipes
* e.g. How to delete multiple values from a vector?
    * https://stackoverflow.com/questions/9665984/how-to-delete-multiple-values-from-a-vector
    * https://stackoverflow.com/

# Plots

Graphically plotting/charting data

## Introduction to plots

```{r}
#R has powerful data plotting abilities
hist(runif(50))
plot(runif(50))
plot(runif(50), runif(50)) #x,y scatter plot
boxplot(runif(50))
demo(graphics)
```

## Exercise: plots (10 minutes)

>Create a matrix “randomdata” with 50 rows and 10 columns using "replicate(10,rnorm(50))"  
>Plot the first and second columns of the matrix in a scatter plot  
>Generate a vector “randommeans” with the means of all columns of randomdata. Also generate a vector “randomsds” with the standard deviations of all columns of randomdata  
>Generate a boxplot of the first and second columns of randomdata

### Exercise solution (5 minutes)
```{r, fig.height=3,fig.width=9}
# Make random (normal!) data ----
randomdata1 <- matrix(rnorm(500),nrow=50,ncol=10)
randomdata2 <- replicate(10,rnorm(50)) 
#^ is a wrapper for sapply(1:10,function(X) rnorm(50))

temp <- list()
for(i in 1:10) {
  temp[[i]] <- rnorm(50)
}
randomdata3 <- do.call("cbind",temp)

dim(randomdata1)
str(randomdata2)
str(randomdata3)


# step 2 ----
default_par <- par(no.readonly=T)
par(mfrow=c(1,3),mar=c(3,3,1,1),mgp=c(2,1,0))
plot(randomdata1[,1:2])
plot(x=randomdata1[,1],y=randomdata1[,2])
rd4 <- as.data.frame(randomdata1)
plot(V2~V1,data=rd4)

par(default_par)

# step 3 ----
rmeans <- colMeans(randomdata1)
rmeans
rsd <- apply(randomdata1,MARGIN=2,FUN=sd)
rsd 

```


```{r}
# step 4 ----
boxplot(randomdata1[,1:2])
boxplot(list("First column"=randomdata1[,1],
             "Le Deux"=randomdata1[,2]))
boxplot(randomdata1)


hist(randomdata1[,1])
hist(randomdata1)

```

***

# Assignment 4 notes
* ~~Do yourself and I a favour by submitted part 1 as an R script.~~  I've read the assignment more closely, and it would be easier for me (and probably you) if you just submitted a text file with each answer (the RegEx or sentence response) on a line, so that I can easily copy and paste into my test code.  Still best for you to write the text file in RStudio so that you know it won't be converted to rich text, and you can easily test your RegEx.  
* For part 2, remember that a FASTA file is a text file where each sequence is preceded by a header line that starts with `>`.  You can use `readLines` to read in the file as a character vector, then loop over it to either start a new sequence if you find `>` else append the line to the current sequence string.  
* To clarify what part of each sequence you should be using: Each sequence in the FASTA file includes the preceding and following 1k base pairs.  All we want you to store in the dataframe is the first 1000 base pairs of the sequence, and the last 1000 base pairs of the sequence.  Don't worry about finding the ORF start codon or anything like that, just paste the whole sequence together, then grab the substrings containing the first and last 1000 bp and paste them together in that dataframe.  **Note that when looking for the transcription factor binding sites, you should only look in the first 1000 bp, so you'll have to substring the sequence again.**  
* Note that the final question wants a percentage *per motif* rather than per transcription factor in the `yeast_tf_motifs.txt` file.  While some of the motifs in that file are ascribed to the same transcription factor, we expect a value per motif.  

****