<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Module 5</title>

<script src="site_libs/header-attrs-2.7/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-sm-12 col-md-4 col-lg-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-sm-12 col-md-8 col-lg-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">MMG1002H</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Module 1</a>
</li>
<li>
  <a href="mod2.html">Module 2</a>
</li>
<li>
  <a href="mod3.html">Module 3</a>
</li>
<li>
  <a href="mod4.html">Module 4</a>
</li>
<li>
  <a href="mod5.html">Module 5</a>
</li>
<li>
  <a href="mod6.html">Module 6</a>
</li>
<li>
  <a href="mod7.html">Module 7</a>
</li>
<li>
  <a href="mod8.html">Module 8</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Module 5</h1>

</div>


<blockquote>
<p>Updated 2021/04/30</p>
</blockquote>
<div id="assignment-2-notes" class="section level1">
<h1>Assignment 2 notes</h1>
<div id="part-3-substring" class="section level2">
<h2>Part 3 substring:</h2>
<p>Hard-coding indices to find the last word in the string is not robust coding practice. Here’s two better ways of doing that:</p>
<pre class="r"><code>string = &quot;This is a kind of a long string&quot;

substr(string,26,31) # not good</code></pre>
<pre><code>## [1] &quot;string&quot;</code></pre>
<pre class="r"><code>substr(string,nchar(string)-5,nchar(string)) # still not good</code></pre>
<pre><code>## [1] &quot;string&quot;</code></pre>
<pre class="r"><code>spaces &lt;- gregexpr(&quot; &quot;,string)[[1]]
substr(string,spaces[length(spaces)] + 1,nchar(string))</code></pre>
<pre><code>## [1] &quot;string&quot;</code></pre>
<pre class="r"><code># good, but uses a function we don&#39;t know yet.

strsplit(string,&quot; &quot;)[[1]][length(strsplit(string,&quot; &quot;)[[1]])]</code></pre>
<pre><code>## [1] &quot;string&quot;</code></pre>
<pre class="r"><code># easier, doesn&#39;t use substr (but substr is a dumb way to handle this anyway).</code></pre>
</div>
<div id="part-4" class="section level2">
<h2>Part 4:</h2>
<p>Many ways of doing this. Here are some!</p>
<pre class="r"><code>sequence &lt;- paste(readLines(&quot;assignment2_sequence.txt&quot;)[1:5],collapse=&quot;&quot;)
sequence &lt;- substr(sequence,regexpr(&quot;AUG&quot;,sequence)[[1]],nchar(sequence))
# I really compressed the above, but you all figured out this part pretty well.

aasequence = &quot;&quot;
index=1
while(index &lt; nchar(sequence)) {
  codon &lt;- substr(sequence, index, index + 2)
  if(codon == &quot;AUG&quot;) {
        aasequence = c(aasequence, &quot;M&quot;)
    } else if (codon == &quot;UUC&quot;) {
        aasequence = c(aasequence, &quot;F&quot;)
    } else if (codon == &quot;CUU&quot;) {
        aasequence = c(aasequence, &quot;L&quot;)
    } else if (codon == &quot;CCA&quot;) {
        aasequence = c(aasequence, &quot;P&quot;)
    } else if (codon == &quot;ACC&quot;) {
        aasequence = c(aasequence, &quot;T&quot;)
    } else if (codon == &quot;GAU&quot;) {
        aasequence = c(aasequence, &quot;D&quot;)
    } else if (codon == &quot;CGA&quot;) {
        aasequence = c(aasequence, &quot;R&quot;)
    } else if (codon == &quot;UGA&quot;) {
        aasequence = c(aasequence, &quot;stop&quot;)
    }
  if (codon == &quot;stop&quot;) { break }
  index = index + 3  #this moves the index forward 3 places in the sequence
}
paste(aasequence[-1],collapse=&quot;&quot;) #why -1 here?</code></pre>
<pre><code>## [1] &quot;MDFRPFDRFPFTLTRFTFPFDLDFTRFDFT&quot;</code></pre>
<pre class="r"><code># a shorter way, using the sub command!
aasequence = &quot;&quot;
index=1
while(index &lt; nchar(sequence)) {
  codon &lt;- substr(sequence, index, index + 2)
  codon &lt;- sub(&quot;AUG&quot;, &quot;M&quot;, codon)
  codon &lt;- sub(&quot;UUC&quot;, &quot;F&quot;, codon)
  codon &lt;- sub(&quot;CUU&quot;, &quot;L&quot;, codon)
  codon &lt;- sub(&quot;CCA&quot;, &quot;P&quot;, codon)
  codon &lt;- sub(&quot;ACC&quot;, &quot;T&quot;, codon)
  codon &lt;- sub(&quot;GAU&quot;, &quot;D&quot;, codon)
  codon &lt;- sub(&quot;CGA&quot;, &quot;R&quot;, codon)
  codon &lt;- sub(&quot;UGA&quot;, &quot;stop&quot;, codon)
  if (codon == &quot;stop&quot;) { break }
  aasequence &lt;- c(aasequence,codon)
  index = index + 3  #this moves the index forward 3 places in the sequence
}
paste(aasequence[-1],collapse=&quot;&quot;) #why -1 here?</code></pre>
<pre><code>## [1] &quot;MDFRPFDRFPFTLTRFTFPFDLDFTRFDFT&quot;</code></pre>
<pre class="r"><code># mapping using indices!
aasequence = &quot;&quot;
index=1
while(index &lt; nchar(sequence)) {
  codon &lt;- substr(sequence, index, index + 2)
  ref_aa&lt;-c(AUG=&quot;M&quot;, UUC=&quot;F&quot;, CUU=&quot;L&quot;, CCA=&quot;P&quot;, ACC=&quot;T&quot;, GAU=&quot;D&quot;, CGA=&quot;R&quot;, UGA=&quot;stop&quot;)
  next_aa &lt;- ref_aa[codon]
  if (next_aa == &quot;stop&quot;) { break }
  aasequence &lt;- c(aasequence,next_aa)
  index = index + 3  #this moves the index forward 3 places in the sequence
}
paste(aasequence[-1],collapse=&quot;&quot;) #why -1 here?</code></pre>
<pre><code>## [1] &quot;MDFRPFDRFPFTLTRFTFPFDLDFTRFDFT&quot;</code></pre>
<pre class="r"><code>aasequence3 &lt;- aasequence[-1]

# finally, using switch (which probably isn&#39;t as effective as using indices, but weirder so it&#39;s last)
aasequence = &quot;&quot;
index=1
while(index &lt; nchar(sequence)) {
  codon &lt;- substr(sequence, index, index + 2)
  next_aa &lt;- switch(codon,
                    AUG=&quot;M&quot;,
                    UUC=&quot;F&quot;,
                    CUU=&quot;L&quot;,
                    CCA=&quot;P&quot;,
                    ACC=&quot;T&quot;,
                    GAU=&quot;D&quot;,
                    CGA=&quot;R&quot;,
                    UGA=&quot;stop&quot;)
  if (next_aa == &quot;stop&quot;) { break }
  aasequence &lt;- c(aasequence,next_aa)
  index = index + 3  #this moves the index forward 3 places in the sequence
}
paste(aasequence[-1],collapse=&quot;&quot;) #why -1 here?</code></pre>
<pre><code>## [1] &quot;MDFRPFDRFPFTLTRFTFPFDLDFTRFDFT&quot;</code></pre>
<p>The Book of R: chapter 10.2 and 10.3 (loops), 10.2.3 (apply functions)</p>
<hr />
</div>
</div>
<div id="module-4-review" class="section level1">
<h1>Module 4 review</h1>
<ul>
<li>Factors
<ul>
<li>f = factor(x)</li>
<li>levels(f)</li>
</ul></li>
<li>Files
<ul>
<li>ASCII</li>
<li>write.table</li>
<li>read.table</li>
</ul></li>
<li>Conditional statements
<ul>
<li>if, if else if, if else if else</li>
</ul></li>
</ul>
<hr />
<div id="exercise-conditions-10-minutes" class="section level2">
<h2>Exercise: conditions (10 minutes)</h2>
<blockquote>
<p>Write conditional statement code that evaluates an amino acid’s physicochemical property and outputs the following, given a single letter amino acid code:</p>
</blockquote>
<pre><code>&quot;Charged&quot; for R, K, D, E 
&quot;Polar&quot; for Q, N, H, S, T, Y, C
&quot;Amphipathic&quot; for W, Y, M
&quot;Hydrophobic&quot; for A, I, L, M, F, V, P, G</code></pre>
<div id="exercise-solution-5-minutes" class="section level3">
<h3>Exercise solution (5 minutes)</h3>
<pre class="r"><code>AAmap &lt;- list(
  Charged=c(&quot;R&quot;,&quot;K&quot;,&quot;D&quot;,&quot;E&quot;),
  Polar=c(&quot;Q&quot;,&quot;N&quot;,&quot;H&quot;,&quot;S&quot;,&quot;T&quot;,&quot;Y&quot;,&quot;C&quot;),
  Amphipathic=c(&quot;W&quot;,&quot;Y&quot;,&quot;M&quot;),
  Hydrophobic=c(&quot;A&quot;,&quot;I&quot;,&quot;L&quot;,&quot;M&quot;,&quot;F&quot;,&quot;V&quot;,&quot;P&quot;,&quot;G&quot;)
)

AA &lt;- &quot;G&quot;


if (AA %in% c(&quot;R&quot;,&quot;K&quot;,&quot;D&quot;,&quot;E&quot;)) {
  property &lt;- &quot;Charged&quot;
} else if (AA %in% AAmap$Polar) {
  property &lt;- &quot;Polar&quot;
} else if (AA %in% AAmap[[&quot;Amphipathic&quot;]]) {
  propery &lt;- &quot;Amphipathic&quot;
} else if ( AA %in% AAmap$Hydrophobic) {
  property &lt;- names(AAmap)[4] # this is just a convoluted way of doing the same as above.
} else {
  stop(paste(AA,&quot;is not a recognized amino acid.&quot;)) 
  # always good to have an &quot;else&quot; in case none of the elseifs catch your input.
}
property</code></pre>
<pre><code>## [1] &quot;Hydrophobic&quot;</code></pre>
<pre class="r"><code># aside 
1 %in% 1:5</code></pre>
<pre><code>## [1] TRUE</code></pre>
<pre class="r"><code>any(1 == 1:5)</code></pre>
<pre><code>## [1] TRUE</code></pre>
<pre class="r"><code>{
  print(&quot;hi&quot;)
  print(&quot;you&quot;)
}</code></pre>
<pre><code>## [1] &quot;hi&quot;
## [1] &quot;you&quot;</code></pre>
<pre class="r"><code># A more &quot;R-like&quot; way of doing this:
property &lt;- names(AAmap)[sapply(AAmap,function(X) AA %in% X)]
property</code></pre>
<pre><code>## [1] &quot;Hydrophobic&quot;</code></pre>
</div>
</div>
</div>
<div id="loops" class="section level1">
<h1>Loops</h1>
<pre class="r"><code>vectorVariable = c(1,-2,3,-5)

#Let&#39;s find all of the positive numbers
if(vectorVariable[1]&gt;0) {
  print(&quot;Positive number&quot;)
}</code></pre>
<pre><code>## [1] &quot;Positive number&quot;</code></pre>
<pre class="r"><code>if(vectorVariable[2]&gt;0) {
  print(&quot;Positive number&quot;)
}
if(vectorVariable[3]&gt;0) {
  print(&quot;Positive number&quot;)
}</code></pre>
<pre><code>## [1] &quot;Positive number&quot;</code></pre>
<pre class="r"><code>if(vectorVariable[4]&gt;0) {
  print(&quot;Positive number&quot;)
}

#Problems with this approach:
#1. There is a lot of repeated code
#2. What happens if the vector is really long?</code></pre>
<ul>
<li>Loops provide a way of efficiently repeating the same code</li>
<li>Two major kinds of loops: for, while</li>
</ul>
<div id="for-loop-e.g.-find-positive-numbers" class="section level2">
<h2>For loop e.g. find positive numbers</h2>
<pre class="r"><code>#Here is the same problem, but solved using a for loop
vectorVariable = c(1,-2,3,-5)
for (value in vectorVariable) {
  if (value &gt; 0) {
    print(value)
  }
}</code></pre>
<pre><code>## [1] 1
## [1] 3</code></pre>
<pre class="r"><code>#What is happening here?  Let&#39;s &quot;unroll&quot; the loop
vectorVariable = c(1,-2,3,-5)
value = vectorVariable[1]
if (value &gt; 0) {
  print(value)
}</code></pre>
<pre><code>## [1] 1</code></pre>
<pre class="r"><code>value = vectorVariable[2]
if (value &gt; 0) {
  print(value)
}
value = vectorVariable[3]
if (value &gt; 0) {
  print(value)
}</code></pre>
<pre><code>## [1] 3</code></pre>
<pre class="r"><code>value = vectorVariable[4]
if (value &gt; 0) {
  print(value)
}

#Note: &#39;value&#39; is just a variable, so can have any name of a variable

#The for loop automatically repeated the same code inside the loop, automatically updating the value variable to the next element of the vector, so it could &quot;loop&quot; through all the vector elements

#here is an idiomatic way to write a for loop that executes a specific number of times:
sum=0
for (i in 1:100) {
  sum = sum + i;   #note: ; indicates the end of a statement, but it is optional in R
}
print (sum)</code></pre>
<pre><code>## [1] 5050</code></pre>
<pre class="r"><code>#this is the same as sum(1:100) - in this case, R has a &#39;high level&#39; command that does the same thing. Internally, sum() uses a loop to do its job.</code></pre>
</div>
<div id="while-loop" class="section level2">
<h2>While loop</h2>
<ul>
<li>Example: find positive numbers in a vector</li>
</ul>
<pre class="r"><code>vectorVariable = c(1,-2,3,-5)
i=1
while (i &lt;= length(vectorVariable)) {
  if (vectorVariable[i] &gt; 0) {
    print(vectorVariable[i])
  }
  i = i + 1
}</code></pre>
<pre><code>## [1] 1
## [1] 3</code></pre>
<pre class="r"><code>#What is happening here?  Let&#39;s &quot;unroll&quot; the loop
vectorVariable = c(1,-2,3,-5)
i=1  #i is an index to keep track of our position in the vector. i=1 means to start with the first element of the vector.
if (i &lt;= length(vectorVariable)) { #if i is smaller or equal to the vector length, then we can run our code
  if (vectorVariable[i] &gt; 0) {
    print(vectorVariable[i])
  }
  i = i + 1  #move to the next element - i now equal to 2
}</code></pre>
<pre><code>## [1] 1</code></pre>
<pre class="r"><code>if (i &lt;= length(vectorVariable)) { #if i is smaller or equal to the vector length, then we can run our code
  if (vectorVariable[i] &gt; 0) {
    print(vectorVariable[i])
  }
  i = i + 1  #move to the next element - i now equal to 3
}
if (i &lt;= length(vectorVariable)) { #if i is smaller or equal to the vector length, then we can run our code
  if (vectorVariable[i] &gt; 0) {
    print(vectorVariable[i])
  }
  i = i + 1  #move to the next element - i now equal to 4
}</code></pre>
<pre><code>## [1] 3</code></pre>
<pre class="r"><code>if (i &lt;= length(vectorVariable)) { #if i is smaller or equal to the vector length, then we can run our code
  if (vectorVariable[i] &gt; 0) {
    print(vectorVariable[i])
  }
  i = i + 1  #move to the next element - i now equal to 5
}
if (i &lt;= length(vectorVariable)) { #i is no longer smaller than the vector length, so don&#39;t run the code
  if (vectorVariable[i] &gt; 0) {
    print(vectorVariable[i])
  }
  i = i + 1
}


#The while and for loops can do the same thing. Sometimes it is more convenient to write a loop one way or another</code></pre>
</div>
<div id="reasons-to-write-loops" class="section level2">
<h2>Reasons to write loops</h2>
<ul>
<li>You want to loop over all elements of a vector (the most common case in R)</li>
</ul>
<pre class="r"><code>for (i in vectorVariable) {
  if (i &gt; 0) {
    print(i)
  }
}</code></pre>
<pre><code>## [1] 1
## [1] 3</code></pre>
<ul>
<li>You want to run a certain number of operations</li>
</ul>
<pre class="r"><code>sum=0
for (i in 1:100) {
  sum = sum + i
}
print (sum)</code></pre>
<pre><code>## [1] 5050</code></pre>
<pre class="r"><code>#same as
i=1
sum=0
while (i &lt;= 100) {
  sum = sum + i
  i = i + 1
}
print (sum)</code></pre>
<pre><code>## [1] 5050</code></pre>
<ul>
<li>You want to run code until a certain condition occurs e.g. find the first instance of “cancer” in the genes2phenotype file</li>
</ul>
<pre class="r"><code>vectorVariable = c(1,-2,3,-5)
i=1
while (vectorVariable[i] != 3) {
  print(vectorVariable[i])
  i = i + 1
}</code></pre>
<pre><code>## [1] 1
## [1] -2</code></pre>
</div>
<div id="infinite-loops" class="section level2">
<h2>Infinite loops</h2>
<ul>
<li>Be careful that the loops don’t go on forever</li>
<li>If this happens, you need to stop or interrupt the R code. In RStudio, you interrupt the code from the Session menu</li>
</ul>
<pre class="r"><code>vectorVariable = c(1,-2,3,-5)
i=1
while (i &lt;= length(vectorVariable)) {
  if (vectorVariable[i] &gt; 0) {
    print(vectorVariable[i])
  }
  #i = i + 1  #We never update the loop index, so i will never be greater than the vector length, so the loop will never end
}

# Not actually going to run this one, or the code would never knit!</code></pre>
</div>
<div id="exercise-loops-1-10-minutes" class="section level2">
<h2>Exercise: Loops 1 (10 minutes)</h2>
<blockquote>
<p>Remember to count the amino acids in a protein sequence? Let’s do it in a different way here, using loops and named vectors (could also use names in lists)</p>
</blockquote>
<blockquote>
<p>Here again the protein sequence:</p>
</blockquote>
<pre><code>&gt; sp|P04637|P53_HUMAN Cellular tumor antigen p53 OS=Homo sapiens
MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLSPDDIEQWFTEDPGPDEAPRMPEAAPPVAPAPAA
PTPAAPAPAPSWPLSSSVPSQKTYQGSYGFRLGFLHSGTAKSVTCTYSPALNKMFCQLAKTCPVQLWVDSTPPPGTRVR
AMAIYKQSQHMTEVVRRCPHHERCSDSDGLAPPQHLIRVEGNLRVEYLDDRNTFRHSVVVPYEPPEVGSDCTTIHYNYM
CNSSCMGGMNRRPILTIITLEDSSGNLLGRNSFEVRVCACPGRDRRTEEENLRKKGEPHHELPPGSTKRALPNNTSSSP
QPKKKPLDGEYFTLQIRGRERFEMFRELNEALELKDAQAGKEPGGSRAHSSHLKSKKGQSTSRHKKLMFKTEGPDSD</code></pre>
<blockquote>
<p>Use a for loop and the command strplit to get each amino acid of the sequence to count the occurrences.</p>
</blockquote>
<blockquote>
<p>One easy way is to first generate a vector (or list) that has named rows, named for each amino acid, such that, for instance, you keep the number of occurrences of Alanines as occ_aa[“A”]. To start, you have to set all the elements of the vector to zero (R has a built in vector called “LETTERS” that contains the whole alphabet)</p>
</blockquote>
<div id="exercise-solution-5-minutes-1" class="section level3">
<h3>Exercise solution (5 minutes)</h3>
<pre class="r"><code># This exercise isn&#39;t in your version of the notebook, but whatever...
# First I copied that FASTA-formatted sequence into a text file, so we could load it easily.

inputFA &lt;- readLines(&quot;mod5exercise.fa&quot;)


# Now we do it the very un-R-like loop way: 
AAcounts &lt;- integer(length=length(LETTERS))
names(AAcounts) &lt;- LETTERS

for (LINE in inputFA) {
  lineAAs &lt;- strsplit(LINE,&quot;&quot;)[[1]] #the [[1]] is to quickly grab the first (and only) list element
  if (lineAAs[1] == &quot;&gt;&quot;) { next } # skip the header line
  for (AA in lineAAs) {
    AAcounts[AA] &lt;- AAcounts[AA] + 1
  }
}
AAcounts</code></pre>
<pre><code>##  A  B  C  D  E  F  G  H  I  J  K  L  M  N  O  P  Q  R  S  T  U  V  W  X  Y  Z 
## 24  0 10 20 30 11 23 12  8  0 20 32 12 14  0 45 15 26 38 22  0 18  4  0  9  0</code></pre>
<pre class="r"><code># of course, vectorization is way more R-appropriate:
AllTheAAs &lt;- strsplit(inputFA[!grepl(&quot;^&gt;&quot;,inputFA)],&quot;&quot;)
table(unlist(AllTheAAs))</code></pre>
<pre><code>## 
##  A  C  D  E  F  G  H  I  K  L  M  N  P  Q  R  S  T  V  W  Y 
## 24 10 20 30 11 23 12  8 20 32 12 14 45 15 26 38 22 18  4  9</code></pre>
<pre class="r"><code># Are they the same?
all(AAcounts[AAcounts &gt; 0] == table(unlist(AllTheAAs)))</code></pre>
<pre><code>## [1] TRUE</code></pre>
</div>
</div>
<div id="exercise-loops-2-10-minutes" class="section level2">
<h2>Exercise: Loops 2 (10 minutes)</h2>
<blockquote>
<p>Read the “gene_condition_source_id.txt” file dataFrame = read.table(“gene_condition_source_id.txt”, header=TRUE, sep=“, fill=TRUE, comment.char =”“, quote =”")</p>
</blockquote>
<blockquote>
<p>Using a for loop, loop over all the rows of “gene_condition_source_id.txt” and find all diseases linked to TP53</p>
</blockquote>
<blockquote>
<p>Do the same thing, but first read in the file again, but set “stringsAsFactors=F” as an argument to read.table()</p>
</blockquote>
<div id="exercise-solution-5-minutes-2" class="section level3">
<h3>Exercise solution (5 minutes)</h3>
<pre class="r"><code>db &lt;- read.table(&quot;D:/Dropbox/TA_MMG1002H_2021/Week1/gene_condition_source_id.txt&quot;,
                 header=T,sep=&quot;\t&quot;,quote=&quot;&quot;,comment.char=&quot;&quot;)

system.time({
    disease_list1 &lt;- db$DiseaseName[db$AssociatedGenes == &quot;TP53&quot;]
})</code></pre>
<pre><code>##    user  system elapsed 
##       0       0       0</code></pre>
<pre class="r"><code>system.time({
    disease_list &lt;- c()
    for (X in 1:nrow(db)) {
        if (db[X,&quot;AssociatedGenes&quot;] == &quot;TP53&quot;) {
            # disease_list &lt;- append(disease_list,db[X,&quot;DiseaseName&quot;])
            disease_list &lt;- c(disease_list,db[X,&quot;DiseaseName&quot;])
        }
    }
})</code></pre>
<pre><code>##    user  system elapsed 
##    0.08    0.01    0.09</code></pre>
<pre class="r"><code>all(disease_list1 == disease_list)</code></pre>
<pre><code>## [1] TRUE</code></pre>
</div>
</div>
</div>
<div id="vectorization" class="section level1">
<h1>Vectorization</h1>
<p>Key Programming Concept in R (but not in many other programming languages)</p>
<ul>
<li>Key feature of R that helps do powerful things with a small amount of code (high level)</li>
<li>Try to to avoid loops and use vectors because loops tend to be slower than vectors in R</li>
</ul>
<pre class="r"><code>vectorVariable = c(1,-2,3,-4)
vectorVariable</code></pre>
<pre><code>## [1]  1 -2  3 -4</code></pre>
<pre class="r"><code>vectorVariable * 5</code></pre>
<pre><code>## [1]   5 -10  15 -20</code></pre>
<pre class="r"><code>vectorVariable * vectorVariable</code></pre>
<pre><code>## [1]  1  4  9 16</code></pre>
<pre class="r"><code>sum(vectorVariable)</code></pre>
<pre><code>## [1] -2</code></pre>
<pre class="r"><code>#similarly, you can use: prod(), mean(), sd(), min(), max()

stringVariable = c(&quot;first&quot;, &quot;last&quot;)
stringVariable</code></pre>
<pre><code>## [1] &quot;first&quot; &quot;last&quot;</code></pre>
<pre class="r"><code>paste(stringVariable, &quot;name&quot;)</code></pre>
<pre><code>## [1] &quot;first name&quot; &quot;last name&quot;</code></pre>
<pre class="r"><code>paste(stringVariable, c(&quot;name1&quot;, &quot;name2&quot;))  #note how vectorization applies the transformation to all strings in the string vector</code></pre>
<pre><code>## [1] &quot;first name1&quot; &quot;last name2&quot;</code></pre>
<div id="logical-selection-from-a-vector" class="section level2">
<h2>Logical selection from a vector</h2>
<pre class="r"><code>vectorVariable&gt;0</code></pre>
<pre><code>## [1]  TRUE FALSE  TRUE FALSE</code></pre>
<pre class="r"><code>vectorVariable==0</code></pre>
<pre><code>## [1] FALSE FALSE FALSE FALSE</code></pre>
<pre class="r"><code>vectorVariable&lt;0</code></pre>
<pre><code>## [1] FALSE  TRUE FALSE  TRUE</code></pre>
<pre class="r"><code>vectorVariable[c(T,F,T,F)]</code></pre>
<pre><code>## [1] 1 3</code></pre>
<pre class="r"><code>vectorVariable[vectorVariable&gt;0]</code></pre>
<pre><code>## [1] 1 3</code></pre>
<pre class="r"><code>ifelse(vectorVariable&gt;0,&quot;positive number&quot;, &quot;negative number&quot;)</code></pre>
<pre><code>## [1] &quot;positive number&quot; &quot;negative number&quot; &quot;positive number&quot; &quot;negative number&quot;</code></pre>
<pre class="r"><code>#remember: for loop to find positive numbers in a vector
vectorVariable = c(1,-2,3,-5)
for (value in vectorVariable) {
  if (value &gt; 0) {
    print(value)
  }
}</code></pre>
<pre><code>## [1] 1
## [1] 3</code></pre>
<pre class="r"><code>#the vectorization way of doing this is:
vectorVariable[vectorVariable&gt;0]</code></pre>
<pre><code>## [1] 1 3</code></pre>
<pre class="r"><code>#much easier!</code></pre>
</div>
<div id="vectorization-with-data-frames" class="section level2">
<h2>Vectorization with data frames</h2>
<pre class="r"><code>name = c(&quot;patient1&quot;, &quot;patient2&quot;, &quot;patient3&quot;)
name</code></pre>
<pre><code>## [1] &quot;patient1&quot; &quot;patient2&quot; &quot;patient3&quot;</code></pre>
<pre class="r"><code>age = c(46, 49, 50)
age</code></pre>
<pre><code>## [1] 46 49 50</code></pre>
<pre class="r"><code>smoker = c(T, F, T)
smoker</code></pre>
<pre><code>## [1]  TRUE FALSE  TRUE</code></pre>
<pre class="r"><code>patientRecords = data.frame(name, age, smoker, stringsAsFactors=F)

#What is going on here?
patientRecords$age&gt;48</code></pre>
<pre><code>## [1] FALSE  TRUE  TRUE</code></pre>
<pre class="r"><code>patientRecords[patientRecords$age&gt;48]  #oops</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["age"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["smoker"],"name":[2],"type":["lgl"],"align":["right"]}],"data":[{"1":"46","2":"TRUE"},{"1":"49","2":"FALSE"},{"1":"50","2":"TRUE"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<pre class="r"><code>patientRecords[ , patientRecords$age&gt;48]  #oops</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["age"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["smoker"],"name":[2],"type":["lgl"],"align":["right"]}],"data":[{"1":"46","2":"TRUE"},{"1":"49","2":"FALSE"},{"1":"50","2":"TRUE"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<pre class="r"><code>patientRecords[patientRecords$age&gt;48 , ]</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["name"],"name":[1],"type":["chr"],"align":["left"]},{"label":["age"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["smoker"],"name":[3],"type":["lgl"],"align":["right"]}],"data":[{"1":"patient2","2":"49","3":"FALSE","_rn_":"2"},{"1":"patient3","2":"50","3":"TRUE","_rn_":"3"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div id="apply-function" class="section level2">
<h2>“apply” function</h2>
<ul>
<li>applies a function to a matrix</li>
</ul>
<pre class="r"><code>#Create a new matrix with number 1 to 30
x = matrix(1:30, nrow=5, ncol=6)
x</code></pre>
<pre><code>##      [,1] [,2] [,3] [,4] [,5] [,6]
## [1,]    1    6   11   16   21   26
## [2,]    2    7   12   17   22   27
## [3,]    3    8   13   18   23   28
## [4,]    4    9   14   19   24   29
## [5,]    5   10   15   20   25   30</code></pre>
<pre class="r"><code>#Sum the rows
apply(x, MARGIN=1, FUN=sum)</code></pre>
<pre><code>## [1]  81  87  93  99 105</code></pre>
<pre class="r"><code>#or equivalently
apply(x, 1, sum)</code></pre>
<pre><code>## [1]  81  87  93  99 105</code></pre>
<pre class="r"><code>#Sum the columns
apply(x, 2, sum)</code></pre>
<pre><code>## [1]  15  40  65  90 115 140</code></pre>
</div>
<div id="lapply" class="section level2">
<h2>lapply()</h2>
<ul>
<li>Similar to apply, but for lists</li>
</ul>
<pre class="r"><code>#Compute the mean for all the columns in a data.frame (which is also a list)
lapply(patientRecords, mean)</code></pre>
<pre><code>## Warning in mean.default(X[[i]], ...): argument is not numeric or logical:
## returning NA</code></pre>
<pre><code>## $name
## [1] NA
## 
## $age
## [1] 48.33333
## 
## $smoker
## [1] 0.6666667</code></pre>
<pre class="r"><code>#You may see lots of other apply functions in R code e.g. mapply, sapply, tapply, vapply</code></pre>
<hr />
</div>
</div>
<div id="assignment-3-notes" class="section level1">
<h1>Assignment 3 notes</h1>
<p>As I said in class, I expect to unzip the submission and see a directory (folder) with your name or some identifying feature in the name, containing all the input data needed to run your R scripts, individual R scripts for each part of the assignments, and for safety’s sake any output files your R scripts are supposed to generate. If they aren’t supposed to generate an output file, no need to include one. Please don’t paste the R console output into a text file and submit that, I need an R script that I can just open and run without errors. Please do run your scripts in a fresh R session to ensure they run without errors. Hint: see below for how to set up RStudio such that when you open a new session, you don’t keep all the junk from your last session (which is dangerous when trying to write reproducible code):<br />
<img src="RStudioOptions.png" alt="Set up RStudio like this!" /><br />
Finally, you should now be skilled enough in R that you know how to write code that will work on data with the same format but different values. In other words, don’t hard-code things in your script. If you’re asked to find the name of the smallest gene, don’t <code>print("the_gene_name")</code>, do something like <code>print(gene_names[which.min(gene_lengths)])</code>, so that if I run the code on data with an even smaller gene, it will still give the correct answer.</p>
<div id="specific-hints-and-clarifications" class="section level2">
<h2>Specific hints and clarifications</h2>
<p>Part 1b: “For exercise, do it both using a for-loop, as well as using an R built-in function.” In other words, we want you to answer this question two ways - once slowly and painfully by writing your own for loop, then check that you got it right by doing it the easy way using existing R functions.<br />
Part 1c: Be careful how you find the W and C strandedness. Part 1d: The “length” column of the database is the ORF length.<br />
Part 2b: The data is reporting phenotypes of gene deletion experiments, and “inviable” is a specific word used to describe a phenotype where the yeast does not survive if the gene is deleted - the gene is essential for viability.<br />
Part 2c: Counting interactions is hard. Don’t just count A of an A-B interaction, but also be careful not to double count situations where A-B and B-A are recorded.<br />
Part 2d: A classic challenge in bioinformatics is when one database contains genes that are missing in another database. If you treat these as zeroes, you’ll skew your calculations (such as mean). Best to either filter them out, or record missing data as <code>NA</code> (see <code>?mean(...,na.rm=T)</code>).</p>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
