<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Module 8</title>

<script src="site_libs/header-attrs-2.7/header-attrs.js"></script>
<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>








<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-sm-12 col-md-4 col-lg-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-sm-12 col-md-8 col-lg-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">MMG1002H</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Module 1</a>
</li>
<li>
  <a href="mod2.html">Module 2</a>
</li>
<li>
  <a href="mod3.html">Module 3</a>
</li>
<li>
  <a href="mod4.html">Module 4</a>
</li>
<li>
  <a href="mod5.html">Module 5</a>
</li>
<li>
  <a href="mod6.html">Module 6</a>
</li>
<li>
  <a href="mod7.html">Module 7</a>
</li>
<li>
  <a href="mod8.html">Module 8</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Module 8</h1>

</div>


<p>The Book of R: Appendix A (packages)</p>
<hr />
<div id="module-7-review" class="section level2">
<h2>Module 7 review</h2>
<ul>
<li>External programs
<ul>
<li>system2</li>
</ul></li>
<li>Introduction to plots
<ul>
<li>plot(runif(50), runif(50)) #x,y scatter plot</li>
<li>boxplot(runif(50))</li>
</ul></li>
</ul>
<hr />
</div>
<div id="packages" class="section level1">
<h1>Packages</h1>
<div id="key-r-concept" class="section level2">
<h2>Key R concept</h2>
<ul>
<li>A place to store a set of functions</li>
<li>Can create a library of functions</li>
<li>Easily import the functions (and data) into R by installing the * package</li>
<li>Packages are available online in package repositories
<ul>
<li>The Comprehensive R Archive Network (CRAN) - <a href="https://cran.r-project.org/" class="uri">https://cran.r-project.org/</a></li>
<li>Bioconductor - <a href="https://www.bioconductor.org/" class="uri">https://www.bioconductor.org/</a></li>
</ul></li>
</ul>
</div>
</div>
<div id="installing-packages" class="section level1">
<h1>Installing packages</h1>
<pre class="r"><code># This downloads the package to your computer
# Permanent download – you only need to do this once
# select a mirror – choose someplace close (faster internet connection), or just the first one
install.packages(&quot;ggplot2&quot;)</code></pre>
<pre class="r"><code># This loads the package so you can use it
# Need to do this each time you restart R
# Sometimes other packages that this package needs will automatically be downloaded – could take time.
library(ggplot2)</code></pre>
<pre><code>## Warning: package &#39;ggplot2&#39; was built under R version 4.0.5</code></pre>
<pre class="r"><code>#now we can use the functions in the package
ggplot(iris, aes(x = Sepal.Length, fill = Species)) + geom_density(alpha = .3)</code></pre>
<p><img src="mod8_files/figure-html/unnamed-chunk-2-1.png" width="672" /></p>
</div>
<div id="package-management" class="section level1">
<h1>Package management</h1>
<pre class="r"><code>#List all installed packages
installed.packages()

#uninstall a package
#e.g. remove.packages(ggplot2)
#remove.packages(pkgs, lib)

#Lists packages that could be updated
old.packages()

#Update all old packages
#R will likely ask you questions about which versions to install</code></pre>
<pre class="r"><code>update.packages()</code></pre>
</div>
<div id="package-documentation" class="section level1">
<h1>Package documentation</h1>
<ul>
<li>Website
<ul>
<li><a href="https://cran.r-project.org/web/packages/kinship2/index.html" class="uri">https://cran.r-project.org/web/packages/kinship2/index.html</a></li>
</ul></li>
<li>Vignette
<ul>
<li>PDF file taking you through an example of the package functionality that you can follow</li>
<li>The first documentation you should review for any package</li>
</ul></li>
</ul>
</div>
<div id="example-kinship2-package" class="section level1">
<h1>Example: kinship2 package</h1>
<pre class="r"><code>install.packages(&quot;kinship2&quot;)</code></pre>
<pre class="r"><code>library(kinship2)</code></pre>
<pre><code>## Warning: package &#39;kinship2&#39; was built under R version 4.0.5</code></pre>
<pre><code>## Loading required package: Matrix</code></pre>
<pre><code>## Loading required package: quadprog</code></pre>
<pre class="r"><code>data(sample.ped) #load sample data
head(sample.ped)</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["ped"],"name":[1],"type":["int"],"align":["right"]},{"label":["id"],"name":[2],"type":["int"],"align":["right"]},{"label":["father"],"name":[3],"type":["int"],"align":["right"]},{"label":["mother"],"name":[4],"type":["int"],"align":["right"]},{"label":["sex"],"name":[5],"type":["int"],"align":["right"]},{"label":["affected"],"name":[6],"type":["int"],"align":["right"]},{"label":["avail"],"name":[7],"type":["int"],"align":["right"]}],"data":[{"1":"1","2":"101","3":"0","4":"0","5":"1","6":"0","7":"0","_rn_":"1"},{"1":"1","2":"102","3":"0","4":"0","5":"2","6":"1","7":"0","_rn_":"2"},{"1":"1","2":"103","3":"135","4":"136","5":"1","6":"1","7":"0","_rn_":"3"},{"1":"1","2":"104","3":"0","4":"0","5":"2","6":"0","7":"0","_rn_":"4"},{"1":"1","2":"105","3":"0","4":"0","5":"1","6":"NA","7":"0","_rn_":"5"},{"1":"1","2":"106","3":"0","4":"0","5":"2","6":"NA","7":"0","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<pre class="r"><code>pedAll = pedigree(id=sample.ped$id, dadid=sample.ped$father, momid=sample.ped$mother, sex=sample.ped$sex, famid=sample.ped$ped)
ped2basic &lt;- pedAll[&#39;2&#39;]
plot(ped2basic)</code></pre>
<p><img src="mod8_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<pre class="r"><code>df2 &lt;- sample.ped[sample.ped$ped==2,]
id2 &lt;- paste(df2$id, c(&quot;John&quot;, &quot;Linda&quot;, &quot;Jack&quot;, &quot;Rachel&quot;, &quot;Joe&quot;, &quot;Deb&quot;, &quot;Lucy&quot;, &quot;Ken&quot;, &quot;Barb&quot;, &quot;Mike&quot;, &quot;Matt&quot;, &quot;Mindy&quot;, &quot;Mark&quot;, &quot;George&quot;), sep=&quot;\n&quot;)
plot(ped2basic, col=ifelse(df2$avail, 2, 1), id=id2)</code></pre>
<p><img src="mod8_files/figure-html/unnamed-chunk-6-2.png" width="672" /></p>
</div>
<div id="bioconductor" class="section level1">
<h1>Bioconductor</h1>
<ul>
<li><p>Over 1000 bioinformatics packages</p></li>
<li><p>Gene expression</p>
<ul>
<li>Import Affymetrix, Illumina, Nimblegen, Agilent, and other platforms. Perform quality assessment, normalization, differential expression, clustering, classification, gene set enrichment, genetical genomics and other workflows for expression, exon, copy number, SNP, methylation and other assays. Access GEO, ArrayExpress, Biomart, UCSC, and other community resources.</li>
</ul></li>
<li><p>High Throughput Assays</p>
<ul>
<li>Import, transform, edit, analyze and visualize flow cytometric, mass spec, HTqPCR, cell-based, and other assays.</li>
</ul></li>
<li><p>Sequence Data</p>
<ul>
<li>Import fasta, fastq, ELAND, MAQ, BWA, Bowtie, BAM, gff, bed, wig, and other sequence formats. Trim, transform, align, and manipulate sequences. Perform quality assessment, ChIP-seq, differential expression, RNA-seq, and other workflows. Access the Sequence Read Archive.</li>
</ul></li>
<li><p>Annotation</p>
<ul>
<li>Use microarray probe, gene, pathway, gene ontology, homology and other annotations. Access GO, KEGG, NCBI, Biomart, UCSC, vendor, and other sources.</li>
</ul></li>
</ul>
</div>
<div id="creating-your-own-package" class="section level1">
<h1>Creating your own package</h1>
<pre class="r"><code>#Install needed packages
install.packages(&quot;devtools&quot;)
library(&quot;devtools&quot;)
install.packages(&quot;roxygen2&quot;)
library(roxygen2)</code></pre>
<pre class="r"><code>#Create a directory for your package
setwd(&quot;parent_directory&quot;)
#Create the directory &#39;mypackage&#39; inside of &#39;parent_directory&#39;
create(&quot;mypackage&quot;)</code></pre>
<ul>
<li>Save functions in a file in:
<ul>
<li>parent_directory/mypackage/R</li>
<li>Example – copy the following files into this folder
<ul>
<li>lengthfunction.r</li>
<li>regexfunction.r</li>
</ul></li>
<li>Can also save each function as a separate file for organizational purposes</li>
</ul></li>
<li>Write documentation
<ul>
<li>Edit DESCRIPTION text file</li>
<li>Write documentation for each function using
<ul>
<li><a href="https://cran.r-project.org/web/packages/roxygen2/" class="uri">https://cran.r-project.org/web/packages/roxygen2/</a></li>
</ul></li>
</ul></li>
</ul>
<pre class="r"><code>#Function documentation example
#&#39; The length of a string (in characters).
#&#39;
#&#39; @param string input character vector
#&#39; @return numeric vector giving number of characters in each element of the
#&#39;   character vector.  Missing strings have missing length.
#&#39; @seealso \code{\link{nchar}} which this function wraps
#&#39; @export
#&#39; @examples
#&#39; str_length(letters)
#&#39; str_length(c(&quot;i&quot;, &quot;like&quot;, &quot;programming&quot;, NA))
str_length &lt;- function(string) {
  string &lt;- check_string(string)
  nc &lt;- nchar(string, allowNA = TRUE)
  is.na(nc) &lt;- is.na(string)
  nc
}</code></pre>
<pre class="r"><code>#note: setwd resets if you run it just one line at a time. You need to run the whole chunk all at once.

#Generate the documentation files
setwd(&quot;./mypackage&quot;)  #(your package directory)
document()
#Install your package
setwd(&quot;..&quot;)   #(go to parent directory)
install(&quot;mypackage&quot;)
#R copies your package to an internal directory
#Load (or activate) your package
library(mypackage)</code></pre>
<pre class="r"><code>#Check that your package is working
help(package=mypackage)
ls(&quot;package:mypackage&quot;)
DNA = &quot;ACACGTACTGATCGGGATATTAG&quot;
allMatchingStrings(&quot;A.A&quot;, DNA)
str_length(DNA)</code></pre>
</div>
<div id="exercise-bioconductor-package-10-minutes" class="section level1">
<h1>Exercise: Bioconductor package (10 minutes)</h1>
<blockquote>
<p>Gene expression analysis example</p>
</blockquote>
<blockquote>
<p>Go to: <a href="http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE13924" class="uri">http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE13924</a></p>
</blockquote>
<blockquote>
<p>Download cel files (2.2MB) (2 arrays)</p>
</blockquote>
<blockquote>
<p>cel files are gene expression microarray scanned image files</p>
</blockquote>
<blockquote>
<p>From “Global transcriptional response of Saccharomyces cerevisiae following the deletion of succinate dehydrogenase” (selected because it is small)</p>
</blockquote>
<blockquote>
<p>Identified using the following search of GEO: <a href="http://www.ncbi.nlm.nih.gov/gds" class="uri">http://www.ncbi.nlm.nih.gov/gds</a> with “GPL90[ACCN] AND gse[ETYP] AND cel[suppFile]” – sort from smallest to largest, download the smallest experiment (2 arrays)</p>
</blockquote>
<blockquote>
<p>Install the affy package</p>
</blockquote>
<pre class="r"><code>#Review the following code and run it, command by command

#install the affy package
# source(&quot;http://www.bioconductor.org/biocLite.R&quot;)
# biocLite(&quot;affy&quot;)
# ^ deprecated method
install.packages(&quot;BiocManager&quot;)
BiocManager::install(&quot;affy&quot;)</code></pre>
<pre class="r"><code>#The package contains functions for exploratory oligonucleotide array analysis
library(affy)</code></pre>
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## Loading required package: parallel</code></pre>
<pre><code>## 
## Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:parallel&#39;:
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     anyDuplicated, append, as.data.frame, basename, cbind, colnames,
##     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
##     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
##     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
##     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
##     union, unique, unsplit, which.max, which.min</code></pre>
<pre><code>## Loading required package: Biobase</code></pre>
<pre><code>## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     &#39;browseVignettes()&#39;. To cite Bioconductor, see
##     &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.</code></pre>
<pre class="r"><code>#loads all .cel files in a directory
Data &lt;- ReadAffy(&quot;../data/GSE13924_RAW/GSM350771.CEL.gz&quot;,
                 &quot;../data/GSE13924_RAW/GSM350772.CEL.gz&quot;)
Data</code></pre>
<pre><code>## installing the source package &#39;ygs98cdf&#39;</code></pre>
<pre><code>## Warning: replacing previous import &#39;AnnotationDbi::tail&#39; by &#39;utils::tail&#39; when
## loading &#39;ygs98cdf&#39;</code></pre>
<pre><code>## Warning: replacing previous import &#39;AnnotationDbi::head&#39; by &#39;utils::head&#39; when
## loading &#39;ygs98cdf&#39;</code></pre>
<pre><code>## </code></pre>
<pre><code>## AffyBatch object
## size of arrays=534x534 features (18 kb)
## cdf=YG_S98 (9335 affyids)
## number of samples=2
## number of genes=9335
## annotation=ygs98
## notes=</code></pre>
<pre class="r"><code>#what do the data distributions look like?
boxplot(Data)</code></pre>
<p><img src="mod8_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>
<pre class="r"><code>hist(Data[,1:2])</code></pre>
<p><img src="mod8_files/figure-html/unnamed-chunk-14-2.png" width="672" /></p>
<pre class="r"><code>#Check out what the cel images look like
par(mfrow=c(1,2),mar=c(1,1,1,1))
image(Data)</code></pre>
<p><img src="mod8_files/figure-html/unnamed-chunk-15-1.png" width="672" /></p>
<pre class="r"><code>#perform probe quality analysis
deg=AffyRNAdeg(Data)
names(deg)</code></pre>
<pre><code>## [1] &quot;N&quot;               &quot;sample.names&quot;    &quot;means.by.number&quot; &quot;ses&quot;            
## [5] &quot;slope&quot;           &quot;pvalue&quot;</code></pre>
<pre class="r"><code>str(deg)</code></pre>
<pre><code>## List of 6
##  $ N              : int 7835
##  $ sample.names   : chr [1:2] &quot;GSM350771.CEL.gz&quot; &quot;GSM350772.CEL.gz&quot;
##  $ means.by.number: num [1:2, 1:16] 6.87 6.26 6.88 6.26 6.86 ...
##  $ ses            : num [1:2, 1:16] 0.019 0.0174 0.0188 0.0173 0.0184 ...
##  $ slope          : num [1:2] -0.00256 -0.06356
##  $ pvalue         : num [1:2] 0.984 0.542</code></pre>
<pre class="r"><code>summaryAffyRNAdeg(deg)</code></pre>
<pre><code>##        GSM350771.CEL.gz GSM350772.CEL.gz
## slope          -0.00256          -0.0636
## pvalue          0.98400           0.5420</code></pre>
<pre class="r"><code>plotAffyRNAdeg(deg)</code></pre>
<p><img src="mod8_files/figure-html/unnamed-chunk-16-1.png" width="672" /></p>
<pre class="r"><code>#Correct background and normalize, log2 values result
eset &lt;- rma(Data)</code></pre>
<pre><code>## Background correcting
## Normalizing
## Calculating Expression</code></pre>
<pre class="r"><code>eset</code></pre>
<pre><code>## ExpressionSet (storageMode: lockedEnvironment)
## assayData: 9335 features, 2 samples 
##   element names: exprs 
## protocolData
##   sampleNames: GSM350771.CEL.gz GSM350772.CEL.gz
##   varLabels: ScanDate
##   varMetadata: labelDescription
## phenoData
##   sampleNames: GSM350771.CEL.gz GSM350772.CEL.gz
##   varLabels: sample
##   varMetadata: labelDescription
## featureData: none
## experimentData: use &#39;experimentData(object)&#39;
## Annotation: ygs98</code></pre>
<pre class="r"><code>#save the results - normalized expression data
write.exprs(eset, file=&quot;mydata.txt&quot;)</code></pre>
<pre class="r"><code>#view the expression data in different ways
par(mfrow=c(1,1))
dim(exprs(eset))</code></pre>
<pre><code>## [1] 9335    2</code></pre>
<pre class="r"><code>boxplot(exprs(eset))</code></pre>
<p><img src="mod8_files/figure-html/unnamed-chunk-19-1.png" width="672" /></p>
<pre class="r"><code>hist(exprs(eset))</code></pre>
<p><img src="mod8_files/figure-html/unnamed-chunk-19-2.png" width="672" /></p>
<pre class="r"><code># featureNames(eset)
exprs(eset[1,c(1,2)])</code></pre>
<pre><code>##          GSM350771.CEL.gz GSM350772.CEL.gz
## 10000_at          1.34052         1.990771</code></pre>
<pre class="r"><code>#compute the log fold change
logFoldChange = exprs(eset[,1])-exprs(eset[,2])

#explore the fold change distributions
hist(logFoldChange)</code></pre>
<p><img src="mod8_files/figure-html/unnamed-chunk-20-1.png" width="672" /></p>
<pre class="r"><code>plot(logFoldChange,pch=&quot;.&quot;)</code></pre>
<p><img src="mod8_files/figure-html/unnamed-chunk-20-2.png" width="672" /></p>
<pre class="r"><code>str(logFoldChange)</code></pre>
<pre><code>##  num [1:9335, 1] -0.65 0.176 -0.506 0.04 -0.339 ...
##  - attr(*, &quot;dimnames&quot;)=List of 2
##   ..$ : chr [1:9335] &quot;10000_at&quot; &quot;10001_at&quot; &quot;10002_i_at&quot; &quot;10003_f_at&quot; ...
##   ..$ : chr &quot;GSM350771.CEL.gz&quot;</code></pre>
<pre class="r"><code>plot(sort(logFoldChange))</code></pre>
<p><img src="mod8_files/figure-html/unnamed-chunk-20-3.png" width="672" /></p>
<pre class="r"><code>#plot heatmap of first 100 probes
heatmap(exprs(eset[1:100,]))</code></pre>
<p><img src="mod8_files/figure-html/unnamed-chunk-21-1.png" width="672" /></p>
<pre class="r"><code>#find the most differentially expressed probes
rownames(logFoldChange)[logFoldChange&gt;4 | logFoldChange &lt; -2]</code></pre>
<pre><code>##  [1] &quot;10194_at&quot;         &quot;10294_at&quot;         &quot;10470_at&quot;         &quot;10825_at&quot;        
##  [5] &quot;10876_at&quot;         &quot;10997_at&quot;         &quot;11133_at&quot;         &quot;11213_at&quot;        
##  [9] &quot;11237_at&quot;         &quot;11297_at&quot;         &quot;3096_f_at&quot;        &quot;3139_f_at&quot;       
## [13] &quot;3310_s_at&quot;        &quot;3326_at&quot;          &quot;3451_f_at&quot;        &quot;3668_f_at&quot;       
## [17] &quot;3807_s_at&quot;        &quot;3849_f_at&quot;        &quot;3853_f_at&quot;        &quot;3864_i_at&quot;       
## [21] &quot;4100_at&quot;          &quot;4114_at&quot;          &quot;4121_at&quot;          &quot;4129_at&quot;         
## [25] &quot;4274_at&quot;          &quot;4506_at&quot;          &quot;4651_at&quot;          &quot;4692_f_at&quot;       
## [29] &quot;4778_at&quot;          &quot;4811_at&quot;          &quot;5089_at&quot;          &quot;5359_at&quot;         
## [33] &quot;5603_at&quot;          &quot;5899_i_at&quot;        &quot;5934_at&quot;          &quot;5953_at&quot;         
## [37] &quot;6556_at&quot;          &quot;6730_at&quot;          &quot;6740_at&quot;          &quot;6764_g_at&quot;       
## [41] &quot;6953_at&quot;          &quot;6969_at&quot;          &quot;7042_at&quot;          &quot;7077_at&quot;         
## [45] &quot;7381_at&quot;          &quot;7458_i_at&quot;        &quot;7513_f_at&quot;        &quot;7544_at&quot;         
## [49] &quot;8192_at&quot;          &quot;8268_at&quot;          &quot;8388_at&quot;          &quot;8513_at&quot;         
## [53] &quot;8557_at&quot;          &quot;8678_at&quot;          &quot;8752_f_at&quot;        &quot;8797_f_at&quot;       
## [57] &quot;8843_at&quot;          &quot;8844_at&quot;          &quot;8850_at&quot;          &quot;8896_at&quot;         
## [61] &quot;9124_at&quot;          &quot;9135_at&quot;          &quot;9196_s_at&quot;        &quot;9198_s_at&quot;       
## [65] &quot;9252_at&quot;          &quot;9259_at&quot;          &quot;9262_at&quot;          &quot;9633_at&quot;         
## [69] &quot;AFFX-18srRnab_at&quot;</code></pre>
<pre class="r"><code>length(rownames(logFoldChange)[logFoldChange&gt;3])</code></pre>
<pre><code>## [1] 54</code></pre>
<pre class="r"><code>#save the results be loaded into another program e.g. a spreadsheet
write.table(logFoldChange, file=&quot;test.txt&quot;, sep=&quot;\t&quot;)</code></pre>
</div>
<div id="exercise-solution-5-minutes" class="section level1">
<h1>Exercise solution (5 minutes)</h1>
</div>
<div id="exercise-gene-expression-and-r-practice-10-minutes" class="section level1">
<h1>Exercise: Gene expression and R practice (10 minutes)</h1>
<blockquote>
<p>Based on the gene expression array data explored in the previous exercise:</p>
</blockquote>
<blockquote>
<p>Plot normalized expression values from both arrays against each other</p>
</blockquote>
<blockquote>
<p>Compare the boxplots of both arrays</p>
</blockquote>
<blockquote>
<p>Compare the histograms of both arrays</p>
</blockquote>
<blockquote>
<p>What percentage of probe sets are under-expressed in the fold change table?</p>
</blockquote>
</div>
<div id="exercise-solution-5-minutes-1" class="section level1">
<h1>Exercise solution (5 minutes)</h1>
<hr />
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
